### ISSUE TITLE
Document support for `dart pub cache preload` and maybe add `dart pub cache export`

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/pub#4081]
#### General info

- Dart 3.2.2 (stable) (Tue Nov 28 11:22:54 2023 +0000) on "windows_x64"
- on windows / "Windows 10 Home" 10.0 (Build 22631)
- locale is en-US

# Problem

When working on a device (eg, a Raspberry Pi) with no internet connection, running `dart pub get` is not an option. Instead, I download packages on my PC then SSH over to the device, copy my Pub cache over, and run `dart pub get --offline` to verify.

Technically, this works, it just takes a while to load because I have to copy over my PC's _entire_ cache. I can run `dart pub cache clean` and `dart pub get` beforehand to make the transfer slightly smaller, but that means any other project won't have the packages it needs when I lose internet again, and if I forget to run `dart pub get` manually each beforehand, that can be bad. 

It would be nice if there were some more offline support for this. For Git repositories, my approach is to push over SSH and use a `dependency_override` to tell Pub to get the repo via path rather than Git URL. But for hosted packages, I need to use the above approach. Maybe there could be a `dart pub cache export <packageName> [version]` that would export a tarball in a format that `dart pub cache preload` could recognize? That way I'd only need to transfer the packages I need instead of my _entire_ cache.

**Comment by sigurdm** [comment]
> I can run dart pub cache clean and dart pub get beforehand to make the transfer slightly smaller, but that means any other project won't have the packages it needs when I lose internet again, and if I forget to run dart pub get manually each beforehand, that can be bad.

You could run 
```
> PUB_CACHE=local_pub_cache dart pub get
```
To get a minimal cache that won't affect your `~/.pub_cache` folder.

**Comment by Levi-Lesches** [comment]
Good idea, I didn't consider changing that value for just one terminal session and exporting that. This makes the process easier. 

Just to clarify then, does `dart pub cache preload` expect each `.tar.gz` passed in the command line to be one individual package, or could I package up my entire Pub cache and pass that in?

**Comment by sigurdm** [comment]
> Just to clarify then, does dart pub cache preload expect each .tar.gz passed in the command line to be one individual package, or could I package up my entire Pub cache and pass that in?

It expects individual packages exactly as they would be downloaded from pub.dev (the hash will be used for validating lock-files)

**Comment by sigurdm** [comment]
You can see how we use it to bootstrap pub.dev: https://github.com/dart-lang/pub-dev/blob/master/tool/pub_get_offline.dart

That is useful if we somehow break pub.dev and want to deploy a fix (hopefully the package buckets on GCS will still be up or all bets are off)

**Comment by jwinarske** [comment]
@Levi-Lesches you can use my script for archive hosted+git and restore for on device:
https://github.com/meta-flutter/meta-flutter/blob/scarthgap/tools/pubspec.py

You could even check the "archive" into git using lfs.

**Comment by Levi-Lesches** [comment]
I'm going to close this and instead plug my package [`pub_cache_server`](https://pub.dev/packages/pub_cache_server) tool that my team has been using for the last year and a half (so, working well). Instead of trying to copy packages manually, `pub_cache_server` turns your own PC's local Pub cache into a Pub server, so all you need to do is change `PUB_HOSTED_URL` on the remote machine and then run `dart pub get` as normal. Works cross-platform, eg, between a Windows dev machine and an offline Linux machine. 

Feel free to re-open if these features are still wanted, but this tool has fixed my use-case above.
